lib/cfp.js

```diff
@@ -0,0 +1,325 @@
+    username: tweet.user.screen_name,
+    tweet: tweet.text,
+    url: `https://twitter.com/${tweet.user.screen_name}/status/${tweet.id_str}`,
+    time: new Date(),
+  },
+  hashtags: await extractTweetHashtags(tweet, options),
+  doc: null,
+  isNew: true,
+});
+
+const formatTweets = async (tweets, options = null) => Promise.all(tweets.map(async (tweet) => {
+  const newTweet = await formatTweet(tweet, options);
+  return newTweet;
+}));
+
+const getBlacklist = async (db = null) => {
+  const useDB = db || getDB();
+  const results = await useDB.collection(COLLECTIONS.BLACKLIST).get();
+  return results.docs.reduce(
+    (acc, doc) => {
+      const nextAcc = { ...acc };
+      const data = doc.data();
+      if (data.hashtag) {
+        nextAcc.hashtags.push(data.hashtag);
+      }
+      if (data.username) {
+        nextAcc.usernames.push(data.username);
+      }
+
+      return nextAcc;
+    },
+    {
+      hashtags: [],
+      usernames: [],
+    },
+  );
+};
+
+const filterBlacklistedTweets = async (formattedTweets, db = null) => {
+  const blacklist = await getBlacklist(db);
+  return formattedTweets.filter((info) => {
+    if (blacklist.usernames.includes(info.tweet.username)) {
+      return false;
+    }
+
+    if (info.hashtags.some((hashtag) => blacklist.hashtags.includes(hashtag))) {
+      return false;
+    }
+
+    return true;
+  });
+};
+
+const saveTweet = async (formattedTweet, db = null) => {
+  const useDB = db || getDB();
+  const collection = useDB.collection(COLLECTIONS.TWEETS);
+  let doc;
+  let isNew = false;
+  const savedTweet = await collection.where('tweetId', '==', formattedTweet.tweet.tweetId).get();
+  if (savedTweet.empty) {
+    doc = await collection.add(formattedTweet.tweet);
+    isNew = true;
+  } else {
+    [doc] = savedTweet.docs;
+  }
+
+  return {
+    ...formattedTweet,
+    ...{
+      doc,
+      isNew,
+    },
+  };
+};
+
+const saveTweets = async (formattedTweets, db = null) => {
+  const newTweets = await Promise.all(formattedTweets.map(async (tweet) => {
+    const newTweet = await saveTweet(tweet, db);
+    return newTweet;
+  }));
+
+  return newTweets.filter((info) => info.isNew);
+};
+
+const saveHashtag = async (hashtag, tweetDocId, db = null) => {
+  const useDB = db || getDB();
+  const collection = useDB.collection(COLLECTIONS.HASHTAGS);
+  const savedHashtag = await collection.where('hashtag', '==', hashtag).get();
+  if (savedHashtag.empty) {
+    await collection.add({
+      hashtag,
+      tweetId: tweetDocId,
+      count: 1,
+    });
+  } else {
+    const [doc] = savedHashtag.docs;
+    await collection.doc(doc.id).update({
+      count: doc.data().count + 1,
+    });
+  }
+};
+
+const saveTweetHashtags = async (formattedTweet, db = null) => {
+  if (!formattedTweet.doc) {
+    throw new Error('You first need to save the tweet');
+  }
+
+  const { hashtags, doc: { id: tweetDocId } } = formattedTweet;
+  await Promise.all(hashtags.map(async (hashtag) => {
```

Comment:
As mentioned elsewhere, I believe the inner `async/await` isn't necessary since you're wrapping in a `Promise.all`:

```js
await Promise.all(hashtags.map((hashtag) => saveHashtag(hashtag, tweetDocId, db)));